Remove static element to value_type map & take id in register_widget instead of proto

raethlein opened the pull request:
## Describe your changes

So far, the `register_widget` function used a static dict to map elements to value types. With the button_group widget we may get the first widget that can either by an `int_array` value or a `trigger_value`. But even without that, we want to switch to a dynamic registration to make widgets more self-contained and open the possibility also for custom components in the future. This reduces the places we have to touch when we add a new element 🙌 
So this PR changes the `register_widget` function so that not the proto but an id has to be passed and the `value_type` is passed directly from the widgets.

## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
  - Added a new Python unit test to ensure that `register_widget` is called for all widgets and that a valid `value_type` is passed.
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

raethlein committed:
Remove static element to value_type map

lukasmasuch commented on lib/streamlit/elements/widgets/button.py:
```diff
@@ -709,14 +709,14 @@ def _download_button(
             deserializer=serde.deserialize,
             serializer=serde.serialize,
             ctx=ctx,
+            value_type="trigger_value",
```
nit: at some point, we maybe should rename this to `bool_trigger_value`

raethlein replied:
Do you think it would make sense to do it now? Or is that a bigger refactoring?

lukasmasuch replied:
Oh, I think the problem is that it's also the name in the proto .. and we still have this "commitment" to not do changes. Maybe add a comment in the proto to rename it. And we should probably discuss about this commitment at offsite ..

raethlein commented:
We don't have to seem a lot of tests around the `register_widget` function itself. I could add a test like this to all widget test suites, e.g. `button_test.py`:

```python
def test_registered_widget_value_type(self):
    """Test that the value_type is registered correctly."""
    st.button("the label", key="foo")
    c = self.get_delta_from_queue().new_element.button

    assert (
        get_script_run_ctx()
        .session_state._state._new_widget_state.widget_metadata[c.id]
        .value_type
        == "trigger_value"
    )
```

I guess it could be helpful, but what do you think? Is that helpful or does it go too deep into the library internals?

lukasmasuch commented on lib/streamlit/runtime/state/widgets.py:
```diff
@@ -37,53 +34,16 @@
     on_change_handler: WidgetCallback | None = None,
     args: WidgetArgs | None = None,
     kwargs: WidgetKwargs | None = None,
+    *,
```
nit: can we move the keyword thing here up above `deserializer`? looks like most uses only use element_proto as positional

raethlein replied:
yep, I will change it 👍

lukasmasuch reviewed (approved):
LGTM 👍 Thanks for refactoring

raethlein committed:
Change keyword-only position

lukasmasuch commented:
> We don't have to seem a lot of tests around the register_widget function itself. I could add a test like this to all widget test suites, e.g. button_test.py

Hmm, maybe we can add a test based on the widget mocks to auto test all widgets. We cannot go as specific, but maybe there are a few more generic tests we could check

raethlein commented:
> Hmm, maybe we can add a test based on the widget mocks to auto test all widgets. We cannot go as specific, but maybe there are a few more generic tests we could check

What do you think about updating the `WIDGET_ELEMENTS` list to 

```python
METADATA: TypeAlias = dict[str, Any]
WIDGET_ELEMENTS: list[tuple[str, ELEMENT_PRODUCER, METADATA]] = [
    # buttons
    ("button", lambda: st.button("Click me"), {"value_type": "trigger_value"}),
...
```
(adding the `METADATA` dict and add a test to check the expectation?

raethlein committed:
Remove element.proto from register_widget

raethlein commented:
@lukasmasuch I have removed the need to pass the proto to the `register_function` and added a test to make sure that `register_widget` is called for all widgets and that a valid `value_type` is passed. These changes reduces two places we had to keep updated when adding a new element.

sfc-gh-lmasuch commented on lib/tests/streamlit/runtime/state/widgets_test.py:
```diff
@@ -587,6 +593,28 @@ def test_widget_id_computation_data_editor(self):
+            widget_func()
+        assert patched_register_widget_from_metadata.call_count == 1
+
+        # ValueFieldName already inclused ArrayValueFieldName, but the check fails
```
inclused -> includes

lukasmasuch commented on lib/tests/streamlit/runtime/state/widgets_test.py:
```diff
@@ -587,6 +593,28 @@ def test_widget_id_computation_data_editor(self):
+            widget_func()
+        assert patched_register_widget_from_metadata.call_count == 1
+
+        # ValueFieldName already inclused ArrayValueFieldName, but the check fails
```
inclused -> includes

raethlein replied:
Oh that was a leftover comment. I removed it completely

lukasmasuch commented on lib/tests/streamlit/runtime/state/widgets_test.py:
```diff
@@ -587,6 +593,28 @@ def test_widget_id_computation_data_editor(self):
+        # in older Python versions
+        assert patched_register_widget_from_metadata.call_args[0][
+            0
+        ].value_type in get_args(ValueFieldName)
```
Would it be possible to also add a check that the value type exist as attribute in the `WidgetState` proto class?

raethlein replied:
Yep! I have just updated the test and added:

```python
# test that the value_type also maps to a protobuf field
assert (
    widget_metadata_arg.value_type
    in WidgetState.DESCRIPTOR.fields_by_name.keys()
)
```

lukasmasuch reviewed (approved):
LGTM 👍
